lib/rust/flexer/src/lib.rs

```diff
@@ -14,26 +15,220 @@
 //! These lexers are capable of lexing any regular grammar, with some extensions to allow working
 //! with context sensitive (e.g. indentation-aware) syntax.
 
-pub use enso_prelude as prelude;
+// TODO [AA] Logging https://github.com/enso-org/ide/blob/main/src/rust/ide/src/model/execution_context/synchronized.rs#L45
 
+use crate::prelude::*;
 
+use lazy_reader::LazyReader;
+use crate::group::Group;
 
 pub mod automata;
-pub mod group;
 pub mod data;
+pub mod group;
 
+#[allow(missing_docs)]
+pub mod prelude {
+    pub use enso_prelude::*;
+}
 
 
-// ===============
-// === Flexer ====
-// ===============
+
+// ==============
+// === Flexer ===
+// ==============
 
 /// The flexer is an engine for generating lexers.
 ///
 /// Akin to flex and other lexer generators, it is given a definition as a series of rules from
 /// which it then generates code for a highly optimised lexer implemented on top of a
 /// [DFA](https://en.wikipedia.org/wiki/Deterministic_finite_automaton).
-pub trait Flexer {
+#[derive(Clone,Debug)]
+pub struct Flexer<Definition,Output,Reader> where Reader:LazyReader {
+    /// The stack of states that are active during lexer execution.
+    pub state_stack: Vec<usize>,
+    /// A reader for the input.
+    pub reader: Reader,
+    /// The current match of the lexer.
+    pub current_match: String,
+    /// The result of the current stage of the DFA.
+    pub status: FlexerStageStatus,
+    /// The tokens that have been lexed.
+    pub tokens: Vec<Output>,
+    /// The definition of the lexer.
+    definition: Definition
+}
+
+impl<Definition,Output,Reader> Deref for Flexer<Definition,Output,Reader>
+where Reader:LazyReader {
+    type Target = Definition;
+    fn deref(&self) -> &Self::Target {
+        &self.definition
+    }
+}
+
+impl<Definition,Output,Reader> DerefMut for Flexer<Definition,Output,Reader>
+where Reader:LazyReader {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.definition
+    }
+}
+
+impl<Definition,Output,Reader> Flexer<Definition,Output,Reader>
+where Definition: FlexerState<Reader>,Reader:LazyReader {
+    /// Creates a new lexer instance.
+    ///
+    /// Please note that the `reader` argument is currently hard-coded for testing purposes. This is
+    /// not the intention for the eventual design.
+    pub fn new(mut reader:Reader) -> Flexer<Definition,Output,Reader> {
+        let mut state_stack = Vec::new();
+        state_stack.reserve(1024);
+        let current_match = String::from("");
+        let status = FlexerStageStatus::Initial;
+        let mut tokens = Vec::new();
+        tokens.reserve(1024);
+        let definition = Definition::new(&mut reader);
+        let initial_state_id = definition.initial_state().id;
```

Comment:
why the initial state is defined in definition? Every lexer should have initial state. It should be part of this sturct imo

Reply:
It needs to be part of the `GroupRegistry` instance so that other groups can use it as a parent. GIven that the `GroupRegistry` lives in the definition, the initial state does too. There are potential workarounds for this, but I think they complicate the API much more.